src/cpp/server/server_cc.cc

```diff
@@ -462,17 +483,25 @@ class Server::CallbackRequest final : public internal::CompletionQueueTag {
           internal::MethodHandler::HandlerParameter(
               call_, &req_->ctx_, req_->request_, req_->request_status_,
               [this] {
-                req_->Reset();
+                // Recycle this request if there aren't too many outstanding.
+                // Note that we don't have to worry about a case where there
+                // are no requests waiting to match for this method since that
+                // is already taken care of when binding a request to a call.
+                if (req_->server_->callback_reqs_outstanding_ <
+                    MAXIMUM_CALLBACK_REQS_OUTSTANDING) {
```

Comment:
Is it better to have a per-method limit rather than a global upper limit?

Reply:
So, it's per-server, not global. And I was thinking that a per-server limit is better because it could be that only 1 method is highly used, so there's no reason to prevent that 1 method from having all the requests allowed for the server. To turn the question around, under what circumstance would a per-method limit be better?

Reply:
I was thinking whether we want to keep the same number of requests (and memory overhead) around for a server no matter how many methods it exposes. For servers exposing very few methods, I was thinking how we can decrease the number of live requests and the memory overhead.

On the other hand, I am not sure whether the naming is precise here. In my understanding, outstanding RPCs are those being processed by the server at any time. callback_reqs_outstanding_ covers both started RPCs and those not-yet-matched free requests. MAXIMUM_CALLBACK_REQS_OUTSTANDING mainly constraints the capacity of the freelist, and I am wondering whether 30,000 is too large for general cases.

I agree in the circumstance you described, per-method limit is not a good option. But can we use a smaller freelist to reduce the memory overhead, and check whether any RPC for a specific method is pending before deleting the request?

Reply:
No, this is not really intended as a freelist size. It was originally inspired by the Stubby freelist code but it went in its own direction after that. `callback_reqs_outstanding_` counts both started RPCs and RPCs that are not yet matched. It is a soft limit in the sense that if there are already MAXIMUM_CALLBACK_REQS_OUTSTANDING outstanding and a new request comes in that is the last unmatched request for its method, another new request will be created. Maybe we need a hard limit also (after which we start rejecting RPCs with errors) but that's another matter.

I would like to allow the size to reduce dynamically as well when we no longer need that many RPCs for the load. I think we need to discuss how to measure the load. One way would be to do a weighted exponential moving average to establish a "target" RPC count and decrease whenever we are above target. I also thought that we might want to increase the count by more than 1 at a time when the load is increasing (perhaps by doubling until we go over a limit, or something like that).